### PR TITLE
support hsm

### PR DESCRIPTION
**Pull request by MaggieNgWu**

### CODE REVIEW

**Review by ywy2090 — APPROVED** [review]

File: build.gradle [hunk 1]
```diff
@@ -193,8 +195,9 @@ dependencies {
     compile ("commons-io:commons-io:${commonsIOVersion}")
     compile ("com.squareup:javapoet:${javapoetVersion}")
     compile ("info.picocli:picocli:${picocliVersion}")
-    compile ("com.webank:key-mini-toolkit:${keyMiniToolkit}")
+    compile ("com.webank:webank-blockchain-java-crypto:${wbjcVersion}")
```
Comment (cyjseagull) [inline build.gradle:198]: Recommend modified to `webankJavaCryptoVersion`

File: sdk-core/src/main/java/org/fisco/bcos/sdk/config/model/AccountConfig.java [hunk 2]
```diff
@@ -38,13 +39,22 @@ public AccountConfig(ConfigProperty configProperty) throws ConfigException {
...
+        this.accountKeyIndex =
+                ConfigProperty.getValue(configProperty.getAccount(), "accountKeyIndex", "");
         if (!this.accountFilePath.equals("")) {
             this.accountFilePath = ConfigProperty.getConfigFilePath(this.accountFilePath);
         }
         checkAccountConfig();
     }
 
     private void checkAccountConfig() throws ConfigException {
+        if (!this.accountKeyIndex.equals("")) {
```
Comment (cyjseagull) [inline sdk-core/src/main/java/org/fisco/bcos/sdk/config/model/AccountConfig.java:51]: Recommend to check this only if the `provider` is `hsm`

File: sdk-core/src/main/java/org/fisco/bcos/sdk/config/model/CryptoProviderConfig.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CryptoProviderConfig {
+    private static Logger logger = LoggerFactory.getLogger(CryptoProviderConfig.class);
+    private String type;
+
+    protected CryptoProviderConfig() {}
+
+    public CryptoProviderConfig(ConfigProperty configProperty) {
```
Comment (cyjseagull) [inline sdk-core/src/main/java/org/fisco/bcos/sdk/config/model/CryptoProviderConfig.java:13]: Recommend the default type is `sw` if the cryptoProvider is not setted.

File: sdk-core/src/test/resources/config-example.toml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 [cryptoMaterial]
-
```
Comment (cyjseagull) [inline sdk-core/src/test/resources/config-example.toml:2]: Please unify the name of `software crypto` and the `hardware crypto`

File: sdk-crypto/build.gradle [hunk 5]
```diff
@@ -5,7 +5,8 @@ plugins {
 dependencies {
     compile project(':sdk-core')
     compile ("org.bouncycastle:bcprov-jdk15on:${bcprovJDK15onVersion}")
-    compile ("com.webank:key-mini-toolkit:${keyMiniToolkit}")
+    compile ("com.webank:webank-blockchain-java-crypto:${wbjcVersion}")
```
Comment (cyjseagull) [inline sdk-crypto/build.gradle:8]: Recommend rename the version to make it more easy to understand.

File: sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/CryptoSuite.java [hunk 6]
```diff
@@ -63,9 +67,46 @@ public CryptoSuite(int cryptoTypeConfig, String hexedPrivateKey) {
      * @param configOption the configuration of account.
      */
     public CryptoSuite(int cryptoTypeConfig, ConfigOption configOption) {
-        this(cryptoTypeConfig);
+        this.cryptoTypeConfig = cryptoTypeConfig;
```
Comment (cyjseagull) [inline sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/CryptoSuite.java:70]: You should modify the `hsm` creation logic in `CryptoSuite(int cryptoConfig)`.

File: sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/CryptoSuite.java [hunk 7]
```diff
@@ -103,6 +142,13 @@ public CryptoSuite(int cryptoTypeConfig) {
         createKeyPair();
     }
 
+    /** Load sdf internal account */
+    public void loadSDFInternalAccount(String accountKeyIndex, String password) {
+        long index = Long.parseLong(accountKeyIndex);
+        SDFSM2KeyPair factory = new SDFSM2KeyPair();
```
Comment (cyjseagull) [inline sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/CryptoSuite.java:148]:
```suggestion
        SDFSM2KeyPair factory = (SDFSM2KeyPair) keyFactory;
```

File: src/test/resources/amop/config-publisher-for-test.toml [hunk 8]
```diff
@@ -1,6 +1,7 @@
 # This is a config file for amop test
 
 [cryptoMaterial]
+provider = "sdf"
```
Comment (cyjseagull) [inline src/test/resources/amop/config-publisher-for-test.toml:4]: Please unify the name of provider.

File: src/test/resources/amop/config-subscriber-for-test.toml [hunk 9]
```diff
@@ -1,6 +1,7 @@
 # This is a config file for amop test
 
 [cryptoMaterial]
+provider = "sdf"
```
Comment (cyjseagull) [inline src/test/resources/amop/config-subscriber-for-test.toml:4]: Please unify the name of provider.

File: src/test/resources/applicationContext-sample.xml [hunk 10]
```diff
@@ -7,6 +7,7 @@
 	<bean id="defaultConfigProperty" class="org.fisco.bcos.sdk.config.model.ConfigProperty">
 		<property name="cryptoMaterial">
 			<map>
+			<entry key="provider" value="sdf" />
```
Comment (cyjseagull) [inline src/test/resources/applicationContext-sample.xml:10]: Please unify the name of provider.

File: src/test/resources/config-example.toml [hunk 11]
```diff
@@ -1,5 +1,5 @@
 [cryptoMaterial]
-
+provider = "sw"                             # sw for software, hsm for hard secure module.
```
Comment (cyjseagull) [inline src/test/resources/config-example.toml:2]: Please unify the name of provider.

File: src/test/resources/config-example2.toml [hunk 12]
```diff
@@ -0,0 +1,65 @@
+[cryptoMaterial]
```
Comment (cyjseagull) [inline src/test/resources/config-example2.toml:1]: Recommend to renama this config to `hsm-config.toml`

File: sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/signature/SDFSM2Signature.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+                        Numeric.cleanHexPrefix(message));
+        logger.info("**** step 1, sdf cal e " + hashResult.hash);
+        if (hashResult.wedprErrorMessage != null && !hashResult.wedprErrorMessage.isEmpty()) {
+            throw new SignatureException(
+                    "Sign with sdf sm2 failed:" + hashResult.wedprErrorMessage);
+        }
+        logger.info("**** step 2, sdf do signature  ");
+        SDFCryptoResult signatureResult;
+        if (keyPair instanceof SDFSM2KeyPair) {
+            SDFSM2KeyPair sdfKeyPair = (SDFSM2KeyPair) keyPair;
```
Comment (cyjseagull) [inline sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/signature/SDFSM2Signature.java:53]: The logic is duplicated， please check this.

File: sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        }
+        this.hexPublicKey =
+                Numeric.getHexKeyWithPrefix(
+                        pkResult.getPublicKey(),
+                        CryptoKeyPair.UNCOMPRESSED_PUBLICKEY_FLAG_STR,
+                        CryptoKeyPair.PUBLIC_KEY_LENGTH_IN_HEX);
+        this.isInternalKey = true;
+    }
+
+    protected SDFSM2KeyPair(long keyIndex) {
```
Comment (cyjseagull) [inline sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java:65]: Please remove this.

File: sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    protected SDFSM2KeyPair(CryptoResult result) {
+        super(result);
+        initSM2KeyPairObject();
+        logger.info(
+                "*** this.hexPrivateKey SDFSM2KeyPair(SDFCryptoResult sm2keyPairInfo) "
+                        + this.hexPrivateKey);
+        this.keyPair = KeyTool.convertHexedStringToKeyPair(this.hexPrivateKey, curveName);
+    }
+
+    protected SDFSM2KeyPair(long keyIndex, String password) {
```
Comment (cyjseagull) [inline sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java:48]: It seems the password is useless, please check again.

File: sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        this.hashImpl = new SDFSM3Hash();
+        this.curveName = CryptoKeyPair.SM2_CURVE_NAME;
+        this.signatureAlgorithm = SM_SIGNATURE_ALGORITHM;
+    }
+
+    @Override
+    public CryptoKeyPair generateKeyPair() {
+        SDFCryptoResult sdfResult = SDF.KeyGen(AlgorithmType.SM2);
+        CryptoResult result = new CryptoResult();
+        result.privateKey = sdfResult.getPrivateKey();
```
Comment (cyjseagull) [inline sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java:92]: Please handle the case that `sdfResult` returns exception.

File: sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java (hunk 15 above)
Comment (MaggieNgWu) [reply sdk-crypto/src/main/java/org/fisco/bcos/sdk/crypto/keypair/SDFSM2KeyPair.java:45]: It is usefull, thanks ^_^

**Review by cyjseagull — APPROVED** [review]

**Pull request merged**